File: src/frontend/src/lib/api-models/canister.ts
import { mapOkResponse } from '@/lib/api-models/error';
import { isNotNil } from '@/lib/nil';
import { fromCandidOpt } from '@/lib/utils';
import type {
  Canister as ApiCanister,
  ListMyCanistersResponse as ApiListMyCanistersResponse,
  ListMyCanistersRequest as ApiListMyCanistersRequest,
  ListUserCanistersResponse as ApiListUserCanistersResponse,
  ListUserCanistersRequest,
} from '@ssn/backend-api';

export type ListMyCanistersResponse = Canister[];
export type ListUserCanistersResponse = {
  canisters: Canister[];
};

export type {
  ListUserCanistersRequest,
  ApiListMyCanistersRequest as ListMyCanistersRequest,
};

export type Canister = {
  id: string;
  principal: string;
  info?: CanisterInfo;
};

export type CanisterInfo = {
  status: CanisterStatus;
  readyForMigration: boolean;
  version: bigint;
  settings: {
    controllers: string[];
    computeAllocation: bigint;
    memoryAllocation: bigint;
    freezingThreshold: bigint;
    reservedCyclesLimit: bigint;
    logVisibility: 'controllers' | 'public' | string[];
    wasmMemoryLimit: bigint;
    wasmMemoryThreshold: bigint;
    environmentVariables: { name: string; value: string }[];
  };
  moduleHash: Uint8Array | null;
  memorySize: bigint;
  memoryMetrics: {
    wasmMemorySize: bigint;
    stableMemorySize: bigint;
    globalMemorySize: bigint;
    wasmBinarySize: bigint;
    customSectionsSize: bigint;
    canisterHistorySize: bigint;
    wasmChunkStoreSize: bigint;
    snapshotsSize: bigint;
  };
  cycles: bigint;
  reservedCycles: bigint;
  idleCyclesBurnedPerDay: bigint;
  queryStats: {
    numCallsTotal: bigint;
    numInstructionsTotal: bigint;
    requestPayloadBytesTotal: bigint;
    responsePayloadBytesTotal: bigint;
  };
};

export enum CanisterStatus {
  Running = 'Running',
  Stopping = 'Stopping',
  Stopped = 'Stopped',
}

export function mapListMyCanistersResponse(
  res: ApiListMyCanistersResponse,
): ListMyCanistersResponse {
  return mapOkResponse(res).map(mapCanisterResponse);
}

export function mapListUserCanistersResponse(
  res: ApiListUserCanistersResponse,
): ListUserCanistersResponse {
  return {
    canisters: mapOkResponse(res).canisters.map(mapCanisterResponse),
  };
}

export function mapCanisterResponse(res: ApiCanister): Canister {
  const canister: Canister = {
    id: res.id,
    principal: res.principal_id,
  };

  const [canisterInfo] = res.info;
  if (isNotNil(canisterInfo)) {
    canister.info = {
      status:
        'Running' in canisterInfo.status
          ? CanisterStatus.Running
          : 'Stopping' in canisterInfo.status
            ? CanisterStatus.Stopping
            : CanisterStatus.Stopped,
      readyForMigration: canisterInfo.ready_for_migration,
      version: canisterInfo.version,
      settings: {
        controllers: canisterInfo.settings.controllers.map(c => c.toText()),
        computeAllocation: canisterInfo.settings.compute_allocation,
        memoryAllocation: canisterInfo.settings.memory_allocation,
        freezingThreshold: canisterInfo.settings.freezing_threshold,
        reservedCyclesLimit: canisterInfo.settings.reserved_cycles_limit,
        logVisibility:
          'Controllers' in canisterInfo.settings.log_visibility
            ? 'controllers'
            : 'Public' in canisterInfo.settings.log_visibility
              ? 'public'
              : canisterInfo.settings.log_visibility.AllowedViewers.map(p =>
                  p.toText(),
                ),
        wasmMemoryLimit: canisterInfo.settings.wasm_memory_limit,
        wasmMemoryThreshold: canisterInfo.settings.wasm_memory_threshold,
        environmentVariables: canisterInfo.settings.environment_variables.map(
          env => ({
            name: env.name,
            value: env.value,
          }),
        ),
      },
      moduleHash: fromCandidOpt(canisterInfo.module_hash),
      memorySize: canisterInfo.memory_size,
      memoryMetrics: {
        wasmMemorySize: canisterInfo.memory_metrics.wasm_memory_size,
        stableMemorySize: canisterInfo.memory_metrics.stable_memory_size,
        globalMemorySize: canisterInfo.memory_metrics.global_memory_size,
        wasmBinarySize: canisterInfo.memory_metrics.wasm_binary_size,
        customSectionsSize: canisterInfo.memory_metrics.custom_sections_size,
        canisterHistorySize: canisterInfo.memory_metrics.canister_history_size,
        wasmChunkStoreSize: canisterInfo.memory_metrics.wasm_chunk_store_size,
        snapshotsSize: canisterInfo.memory_metrics.snapshots_size,
      },
      cycles: canisterInfo.cycles,
      reservedCycles: canisterInfo.reserved_cycles,
      idleCyclesBurnedPerDay: canisterInfo.idle_cycles_burned_per_day,
      queryStats: {
        numCallsTotal: canisterInfo.query_stats.num_calls_total,
        numInstructionsTotal: canisterInfo.query_stats.num_instructions_total,
        requestPayloadBytesTotal:
          canisterInfo.query_stats.request_payload_bytes_total,
        responsePayloadBytesTotal:
          canisterInfo.query_stats.response_payload_bytes_total,
      },
    };
  }

  return canister;
}

File: src/backend/src/mapping/canister.rs
use crate::{
    data,
    dto::{
        Canister, CanisterInfo, CanisterSettings, CanisterState, CanisterStatus,
        EnvironmentVariable, LogVisibility, MemoryMetrics, QueryStats,
    },
};
use canister_utils::Uuid;
use ic_cdk::management_canister::{self, CanisterStatusResult, CanisterStatusType};

pub fn map_canister_response(
    id: &Uuid,
    canister: &data::Canister,
    state: CanisterState,
) -> Canister {
    Canister {
        id: id.to_string(),
        principal_id: canister.principal.to_string(),
        name: canister.name.clone(),
        state,
    }
}

pub fn map_canister_info(info: CanisterStatusResult) -> CanisterInfo {
    CanisterInfo {
        status: match info.status {
            CanisterStatusType::Running => CanisterStatus::Running,
            CanisterStatusType::Stopping => CanisterStatus::Stopping,
            CanisterStatusType::Stopped => CanisterStatus::Stopped,
        },
        ready_for_migration: info.ready_for_migration,
        version: info.version,
        settings: CanisterSettings {
            controllers: info.settings.controllers,
            compute_allocation: info.settings.compute_allocation,
            memory_allocation: info.settings.memory_allocation,
            freezing_threshold: info.settings.freezing_threshold,
            reserved_cycles_limit: info.settings.reserved_cycles_limit,
            log_visibility: match info.settings.log_visibility {
                management_canister::LogVisibility::Controllers => LogVisibility::Controllers,
                management_canister::LogVisibility::Public => LogVisibility::Public,
                management_canister::LogVisibility::AllowedViewers(principals) => {
                    LogVisibility::AllowedViewers(principals)
                }
            },
            wasm_memory_limit: info.settings.wasm_memory_limit,
            wasm_memory_threshold: info.settings.wasm_memory_threshold,
            environment_variables: info
                .settings
                .environment_variables
                .into_iter()
                .map(|environment_variable| EnvironmentVariable {
                    name: environment_variable.name,
                    value: environment_variable.value,
                })
                .collect(),
        },
        module_hash: info.module_hash,
        memory_size: info.memory_size,
        memory_metrics: MemoryMetrics {
            wasm_memory_size: info.memory_metrics.wasm_memory_size,
            stable_memory_size: info.memory_metrics.stable_memory_size,
            global_memory_size: info.memory_metrics.global_memory_size,
            wasm_binary_size: info.memory_metrics.wasm_binary_size,
            custom_sections_size: info.memory_metrics.custom_sections_size,
            canister_history_size: info.memory_metrics.canister_history_size,
            wasm_chunk_store_size: info.memory_metrics.wasm_chunk_store_size,
            snapshots_size: info.memory_metrics.snapshots_size,
        },
        cycles: info.cycles,
        reserved_cycles: info.reserved_cycles,
        idle_cycles_burned_per_day: info.idle_cycles_burned_per_day,
        query_stats: QueryStats {
            num_calls_total: info.query_stats.num_calls_total,
            num_instructions_total: info.query_stats.num_instructions_total,
            request_payload_bytes_total: info.query_stats.request_payload_bytes_total,
            response_payload_bytes_total: info.query_stats.response_payload_bytes_total,
        },
    }
}
